source/server/api_listener_impl.cc

```diff
@@ -0,0 +1,95 @@
+#include "server/api_listener_impl.h"
+
+#include "envoy/api/v2/lds.pb.h"
+#include "envoy/api/v2/listener/listener.pb.h"
+#include "envoy/stats/scope.h"
+
+#include "common/http/conn_manager_impl.h"
+#include "common/network/resolver_impl.h"
+#include "common/protobuf/utility.h"
+
+#include "server/drain_manager_impl.h"
+#include "server/listener_manager_impl.h"
+
+#include "extensions/filters/network/http_connection_manager/config.h"
+
+namespace Envoy {
+namespace Server {
+
+ApiListenerImpl::ApiListenerImpl(const envoy::config::listener::v3alpha::Listener& config,
+                                 ListenerManagerImpl& parent, const std::string& name,
+                                 ProtobufMessage::ValidationVisitor& validation_visitor)
+    : config_(config), parent_(parent), name_(name),
+      address_(Network::Address::resolveProtoAddress(config.address())),
+      validation_visitor_(validation_visitor),
+      global_scope_(parent_.server_.stats().createScope("")),
+      listener_scope_(parent_.server_.stats().createScope(fmt::format("listener.api.{}.", name_))),
+      read_callbacks_(SyntheticReadCallbacks(*this)) {}
+
+AccessLog::AccessLogManager& ApiListenerImpl::accessLogManager() {
+  return parent_.server_.accessLogManager();
+}
+Upstream::ClusterManager& ApiListenerImpl::clusterManager() {
+  return parent_.server_.clusterManager();
+}
+Event::Dispatcher& ApiListenerImpl::dispatcher() { return parent_.server_.dispatcher(); }
+Network::DrainDecision& ApiListenerImpl::drainDecision() { return *this; }
+Grpc::Context& ApiListenerImpl::grpcContext() { return parent_.server_.grpcContext(); }
+bool ApiListenerImpl::healthCheckFailed() { return parent_.server_.healthCheckFailed(); }
+Tracing::HttpTracer& ApiListenerImpl::httpTracer() { return httpContext().tracer(); }
+Http::Context& ApiListenerImpl::httpContext() { return parent_.server_.httpContext(); }
+Init::Manager& ApiListenerImpl::initManager() { return parent_.server_.initManager(); }
```

Comment:
drive by from https://github.com/envoyproxy/envoy/pull/9205#discussion_r363866697
I am thinking about adding a FilterChainFactoryContext without a ListenerImpl and InitManager is the first concern.
While looking at your impl here, I feel this is probably a glitch.
You are reusing the server's init manager. It is guaranteed to be initialized after worker thread started(search for ListenerManagerImpl::workers_started_)
So if api listener is added through LDS *after* server is initialized the init manager the init manager here doesn't HttpConnManager.

Reply:
Right now the ApiListener can only be created during initial bootstrap so that is why we are only using the server's init manager. I plan on adding dynamic ApiListeners in a subsequent PR where, as you say, the init manager will have to change to a local one like in the ListenerImpl.